Mono.Debugging/Mono.Debugging.Client/StackFrame.cs

```diff
@@ -122,63 +123,98 @@ internal void Attach (DebuggerSession debugSession)
-								EventHandler updated = (s, e) => {
-									manualReset.Set ();
-								};
-								args [n].ValueChanged += updated;
-								if (args [n].IsEvaluating)
-									if (!manualReset.WaitOne (2000))
-										session.OnDebuggerOutput (true, "Timeout evaluating parameters for StackFrame.");
-								args [n].ValueChanged -= updated;
-							}
-							if (n > 0)
-								methodNameBuilder.Append (", ");
-							if (options.StackFrameFormat.ParameterTypes) {
-								methodNameBuilder.Append (args [n].TypeName);
-								if (options.StackFrameFormat.ParameterNames)
-									methodNameBuilder.Append (" ");
+		public virtual string GetFullStackFrameText (EvaluationOptions options)
+		{
+			using (var cts = new CancellationTokenSource (2000))
+				return GetFullStackFrameTextAsync (options, false, cts.Token).GetAwaiter ().GetResult ();
+		}
+
+		public Task<string> GetFullStackFrameTextAsync (CancellationToken cancellationToken = default (CancellationToken))
+		{
+			return GetFullStackFrameTextAsync (session.EvaluationOptions, true, cancellationToken);
+		}
+
+		public virtual Task<string> GetFullStackFrameTextAsync (EvaluationOptions options, CancellationToken cancellationToken = default (CancellationToken))
+		{
+			return GetFullStackFrameTextAsync (options, true, cancellationToken);
+		}
+
+		async Task<string> GetFullStackFrameTextAsync (EvaluationOptions options, bool doAsync, CancellationToken cancellationToken)
+		{
+			var methodNameBuilder = new StringBuilder (SourceLocation.MethodName);
+
+			options = options.Clone ();
+			if (options.StackFrameFormat.ParameterValues) {
+				options.AllowMethodEvaluation = true;
+				options.AllowToStringCalls = true;
+				options.AllowTargetInvoke = true;
+			} else {
+				options.AllowMethodEvaluation = false;
+				options.AllowToStringCalls = false;
+				options.AllowTargetInvoke = false;
+			}
+
+			var args = GetParameters (options);
+
+			//MethodName starting with "["... it's something like [ExternalCode]
+			if (!SourceLocation.MethodName.StartsWith ("[", StringComparison.Ordinal)) {
+				if (options.StackFrameFormat.Module && !string.IsNullOrEmpty (FullModuleName))
+					methodNameBuilder.Insert (0, System.IO.Path.GetFileName (FullModuleName) + "!");
+
+				if (options.StackFrameFormat.ParameterTypes || options.StackFrameFormat.ParameterNames || options.StackFrameFormat.ParameterValues) {
+					methodNameBuilder.Append ("(");
+					for (int n = 0; n < args.Length; n++) {
+						if (args[n].IsEvaluating) {
+							var tcs = new TaskCompletionSource<bool> ();
+							EventHandler updated = (s, e) => {
+								tcs.TrySetResult (true);
+							};
+							args[n].ValueChanged += updated;
+							try {
+								using (var registration = cancellationToken.Register (() => tcs.TrySetCanceled ())) {
+									if (args[n].IsEvaluating) {
+										if (doAsync) {
+											await tcs.Task.ConfigureAwait (false);
+										} else {
+											tcs.Task.Wait (cancellationToken);
```

Comment:
Should we also be using the evaluation timeout here and above by linking a new token source with the one passed in?

Reply:
Hmmm, probably.

I also discovered that there's a way to cancel outstanding requests (theoretically if the backend supports it, so I need to check the protocolClient's capabilities).

Reply:
> Should we also be using the evaluation timeout here and above by linking a new token source with the one passed in?

Actually, you know what? Nah.

Why should we cancel it if it takes longer than a timeout when our UI layer can cancel a token?